README.md
# RPi_mcp3008
RPi_mcp3008 is a library to listen to the MCP3008 A/D converter chip with a RPi.
This library implements the example communication protocol described in the [datasheet](https://www.adafruit.com/datasheets/MCP3008.pdf).


Communication is made through RPi SPI port using [SpiDev](https://github.com/doceme/py-spidev)

## Wiring
Connect the SPI data cables in the tables bellow. Choose either CE0# or CE1# to connect to CS.

### RPi SPI GPIOs

| RPi GPIO  | Mode |
|-----------|:-----|
| GPIO 07   | CE1# |
| GPIO 08   | CE0# |
| GPIO 09   | MISO |
| GPIO 10   | MOSI |
| GPIO 11   | SCLK |


### MCP3008 Pinout

| Pin | Description | Pin | Description |
|-----|:------------|:----|:------------|
| 01  |     CH0     | 09  | Vdd - Supply voltage (2.7V - 5.5V) |
| 02  |     CH1     | 10  | Vref - Reference voltage |
| 03  |     CH2     | 11  | AGND - Analog ground |
| 04  |     CH3     | 12  | CLK - SPI Clock (SCLK) |
| 05  |     CH4     | 13  | Dout - Data out (MISO) |
| 06  |     CH5     | 14  | Din - Data in (MOSI) |
| 07  |     CH6     | 15  | CS - Chip select (CE0# or CE1#) |
| 08  |     CH7     | 16  | DGND - Digital ground |

Please check the [Adafruit guide](https://learn.adafruit.com/reading-a-analog-in-and-controlling-audio-volume-with-the-raspberry-pi/connecting-the-cobbler-to-a-mcp3008) on the MCP3008 for more information about wiring


## Usage

RPi_mcp3008 uses the `with` statement to properly handle the SPI bus cleanup.
```python
import mcp3008
with mcp3008.MCP3008() as adc:
    print adc.read([mcp3008.CH0]) # prints raw data [CH0]
```
It's possible instantiate the object normally, but it's necessary to call the close method before terminating the program.
```python
import mcp3008
adc = mcp3008.MCP3008()
print adc.read([mcp3008.CH0]) # prints raw data [CH0]
adc.close()
```
The initialization arguments are `MCP3008(bus=0, device=0)` where:
`MCP3008(X, Y)` will open `/dev/spidev-X.Y`, same as `spidev.SpiDev.open(X, Y)`
Both arguments are optional and have a default value of `0`

### Methods
Currently there are two implemented methods:
```python
def read(self, modes, norm=False):
    '''
    Returns the raw value (0 ... 1024) of the reading.
    The modes argument is a list with the modes of operation to be read (e.g.
    [mcp3008.CH0,mcp3008.Df0]).
    norm is a normalization factor, usually Vref.
    '''
```

```python
def read_all(self, norm=False):
    '''
    Returns a list with the readings of all the modes
    Data Order:
    [DF0, DF1, DF2, DF3, DF4, DF5, DF6, DF7,
     CH0, CH1, CH2, CH3, CH4, CH5, CH6, CH7]
    norm is a normalization factor, usually Vref.
    '''
```
* The `modes` argument must be a list with at least one of 16 modes listed [bellow](## MCP3008 Operation Modes)
* The `norm` argument is a normalization factor that rescales raw data, usually Vref

### Fixed mode
You can also declare the class with a `fixed mode`, which will make the instance callable and always return the value of the listed modes.
Again you can normalize the data with the norm argument when calling the instance.

```python
import mcp3008
with mcp3008.MCP3008.fixed([mcp3008.CH0, mcp3008.DF0]) as adc:
    print adc()     # prints raw data [CH0, DF0]
    print adc(5.2)  # prints normalized data [CH0, DF0]
```

## MCP3008 Operation Modes
MCP3008 has 16 different operation modes:
It can listen to each of the channels individually **Single Ended** or in a pseudo-differential mode **Differential**

| Single Ended | Differential |
|--------------|:-------------|
| CH0  | DF0  (CH0 = IN+; CH1 = IN-) |
| CH1  | DF0  (CH0 = IN-; CH1 = IN+) |
| CH2  | DF0  (CH2 = IN+; CH3 = IN-) |
| CH3  | DF0  (CH2 = IN-; CH3 = IN+) |
| CH4  | DF0  (CH4 = IN+; CH5 = IN-) |
| CH5  | DF0  (CH4 = IN-; CH5 = IN+) |
| CH6  | DF0  (CH6 = IN+; CH7 = IN-) |
| CH7  | DF0  (CH6 = IN-; CH7 = IN+) |

Use the table above as the operation mode when calling `MCP3008.read(modes)` or setting the `MCP3008.fixed(modes)` mode. (e.g. `MCP3008.read([mcp3008.CH0, mcp3008.DF1])`)


CHANGELOG.md
# CHANGELOG

## 0.1a2
* Converted module into package

## 0.1a1
* Fixed python 3 support
* Added `example.py`

## 0.1.0a0
* Initial release


mcp3008/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
RPi_mcp3008 is a library to listen to the MCP3008 A/D converter chip,
as described in the datasheet.
https://www.adafruit.com/datasheets/MCP3008.pdf

Copyright (C) 2015 Luiz Eduardo Amaral <luizamaral306@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import spidev

# Modes Single
CH0 = 8     # single-ended CH0
CH1 = 9     # single-ended CH1
CH2 = 10    # single-ended CH2
CH3 = 11    # single-ended CH3
CH4 = 12    # single-ended CH4
CH5 = 13    # single-ended CH5
CH6 = 14    # single-ended CH6
CH7 = 15    # single-ended CH7
# Modes Diff
DF0 = 0     # differential CH0 = IN+ CH1 = IN-
DF1 = 1     # differential CH0 = IN- CH1 = IN+
DF2 = 2     # differential CH2 = IN+ CH3 = IN-
DF3 = 3     # differential CH2 = IN- CH3 = IN+
DF4 = 4     # differential CH4 = IN+ CH5 = IN-
DF5 = 5     # differential CH4 = IN- CH5 = IN+
DF6 = 6     # differential CH6 = IN+ CH7 = IN-
DF7 = 7     # differential CH6 = IN- CH7 = IN+

RESOLUTION = 1 << 10 # 10 bits resolution

class MCP3008(spidev.SpiDev):
    '''
    Object that listens the MCP3008 in the SPI port of the RPi.
    Connects the object to the specified SPI device.
    The initialization arguments are MCP3008(bus=0, device=0) where:
    MCP3008(X, Y) will open /dev/spidev-X.Y, same as spidev.SpiDev.open(X, Y).
    '''
    def __init__(self, bus=0, device=0):
        self.bus = bus
        self.device = device
        self.open(self.bus, self.device)
        self.modes = False

    def __del__(self):
        self.close()

    def __enter__(self):
        return self

    def __exit__(self, type, value, tb):
        self.close()

    def __repr__(self):
        return 'MCP3008 object at bus {0}, device {1}'.format(self.bus, self.device)

    def __call__(self, norm=False):
        return self.read(self.modes, norm)

    @classmethod
    def fixed(cls, modes, bus=0, device=0):
        '''
        Initializes the class with fixed modes, which turns the instance callable.
        The modes argument is a list with the modes of operation to be read (e.g.
        [mcp3008.CH0,mcp3008.Df0]).
        When calling the instance the object will execute a reading of and return the
        values (e.g. print instance()).
        When calling the instance, you can pass the optional argument norm to
        normalize
        the data (e.g. print instance(5.2)).
        '''
        instance = cls(bus, device)
        instance.modes = modes
        return instance

    def _read_single(self, mode):
        '''
        Returns the value of a single mode reading
        '''
        if not 0 <= mode <= 15:
            raise IndexError('Outside the channels scope, please use: 0, 1 ..., 7')
        request = [0x1, mode << 4, 0x0] # [start bit, configuration, listen space]
        _, byte1, byte2 = self.xfer2(request)
        value = (byte1%4 << 8) + byte2
        return value

    def read(self, modes, norm=False):
        '''
        Returns the raw value (0 ... 1024) of the reading.
        The modes argument is a list with the modes of operation to be read (e.g.
        [mcp3008.CH0,mcp3008.Df0]).
        norm is a normalization factor, usually Vref.
        '''
        reading = []
        for mode in modes:
            reading.append(self._read_single(mode))
        if norm:
            return [float(norm)*value/RESOLUTION for value in reading]
        else:
            return reading

    def read_all(self, norm=False):
        '''
        Returns a list with the readings of all the modes
        Data Order:
        [DF0, DF1, DF2, DF3, DF4, DF5, DF6, DF7,
         CH0, CH1, CH2, CH3, CH4, CH5, CH6, CH7]
        norm is a normalization factor, usually Vref.
        '''
        return self.read(range(16), norm)


mcp3008.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
'''
RPi_mcp3008 is a library to listen to the MCP3008 A/D converter chip,
as described in the datasheet.
https://www.adafruit.com/datasheets/MCP3008.pdf

Copyright (C) 2015 Luiz Eduardo Amaral <luizamaral306@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import spidev

# Modes Single
CH0 = 8     # single-ended CH0
CH1 = 9     # single-ended CH1
CH2 = 10    # single-ended CH2
CH3 = 11    # single-ended CH3
CH4 = 12    # single-ended CH4
CH5 = 13    # single-ended CH5
CH6 = 14    # single-ended CH6
CH7 = 15    # single-ended CH7
# Modes Diff
DF0 = 0     # differential CH0 = IN+ CH1 = IN-
DF1 = 1     # differential CH0 = IN- CH1 = IN+
DF2 = 2     # differential CH2 = IN+ CH3 = IN-
DF3 = 3     # differential CH2 = IN- CH3 = IN+
DF4 = 4     # differential CH4 = IN+ CH5 = IN-
DF5 = 5     # differential CH4 = IN- CH5 = IN+
DF6 = 6     # differential CH6 = IN+ CH7 = IN-
DF7 = 7     # differential CH6 = IN- CH7 = IN+

RESOLUTION = 1 << 10 # 10 bits resolution

class MCP3008(spidev.SpiDev):
    '''
    Object that listens the MCP3008 in the SPI port of the RPi.
    Connects the object to the specified SPI device.
    The initialization arguments are MCP3008(bus=0, device=0) where:
    MCP3008(X, Y) will open /dev/spidev-X.Y, same as spidev.SpiDev.open(X, Y).
    '''
    def __init__(self, bus=0, device=0):
        self.bus = bus
        self.device = device
        self.open(self.bus, self.device)
        self.modes = False

    def __del__(self):
        self.close()

    def __enter__(self):
        return self

    def __exit__(self, type, value, tb):
        self.close()

    def __repr__(self):
        return 'MCP3008 object at bus {0}, device {1}'.format(self.bus, self.device)

    def __call__(self, norm=False):
        return self.read(self.modes, norm)

    @classmethod
    def fixed(cls, modes, bus=0, device=0):
        '''
        Initializes the class with fixed modes, which turns the instance callable.
        The modes argument is a list with the modes of operation to be read (e.g.
        [mcp3008.CH0,mcp3008.Df0]).
        When calling the instance the object will execute a reading of and return the
        values (e.g. print instance()).
        When calling the instance, you can pass the optional argument norm to
        normalize
        the data (e.g. print instance(5.2)).
        '''
        instance = cls(bus, device)
        instance.modes = modes
        return instance

    def _read_single(self, mode):
        '''
        Returns the value of a single mode reading
        '''
        if not 0 <= mode <= 15:
            raise IndexError('Outside the channels scope, please use: 0, 1 ..., 7')
        request = [0x1, mode << 4, 0x0] # [start bit, configuration, listen space]
        _, byte1, byte2 = self.xfer2(request)
        value = (byte1%4 << 8) + byte2
        return value

    def read(self, modes, norm=False):
        '''
        Returns the raw value (0 ... 1024) of the reading.
        The modes argument is a list with the modes of operation to be read (e.g.
        [mcp3008.CH0,mcp3008.Df0]).
        norm is a normalization factor, usually Vref.
        '''
        reading = []
        for mode in modes:
            reading.append(self._read_single(mode))
        if norm:
	    print [value for value in reading]
            return [float(norm)*value/RESOLUTION for value in reading]
        else:
            return reading

    def read_all(self, norm=False):
        '''
        Returns a list with the readings of all the modes
        Data Order:
        [DF0, DF1, DF2, DF3, DF4, DF5, DF6, DF7,
         CH0, CH1, CH2, CH3, CH4, CH5, CH6, CH7]
        norm is a normalization factor, usually Vref.
        '''
        return self.read(range(16), norm)

if __name__ == '__main__':
    with MCP3008() as chip:
        print(chip.read_all(4.35))
        print(chip.read_all())
    with MCP3008.fixed([CH0]) as chip:
        print chip()
        print chip(4.35)
